platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/RunBuildBuildOperationMapper.java

```diff
@@ -0,0 +1,103 @@
+import org.gradle.tooling.internal.protocol.InternalFailure;
+import org.gradle.tooling.internal.protocol.events.InternalOperationFinishedProgressEvent;
+import org.gradle.tooling.internal.protocol.events.InternalOperationStartedProgressEvent;
+import org.gradle.tooling.internal.protocol.events.InternalProgressEvent;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.function.Supplier;
+
+@NonNullApi
+class RunBuildBuildOperationMapper implements BuildOperationMapper<RunBuildBuildOperationType.Details, DefaultBuildBuildDescriptor> {
+
+    private final Supplier<OperationIdentifier> operationIdentifierSupplier;
+
+    public RunBuildBuildOperationMapper(BuildOperationIdFactory idFactory) {
+        this.operationIdentifierSupplier = () -> new OperationIdentifier(idFactory.nextId());
+    }
+
+    @Override
+    public boolean isEnabled(BuildEventSubscriptions subscriptions) {
+        return subscriptions.isRequested(OperationType.PROBLEMS);
+    }
+
+    @Override
+    public Class<RunBuildBuildOperationType.Details> getDetailsType() {
+        return RunBuildBuildOperationType.Details.class;
+    }
+
+    @Override
+    public DefaultBuildBuildDescriptor createDescriptor(RunBuildBuildOperationType.Details details, BuildOperationDescriptor buildOperation, @Nullable OperationIdentifier parent) {
+        OperationIdentifier id = buildOperation.getId();
+        String displayName = buildOperation.getDisplayName();
+        return new DefaultBuildBuildDescriptor(id, displayName, displayName, parent);
+    }
+
+    @Override
+    public InternalOperationStartedProgressEvent createStartedEvent(DefaultBuildBuildDescriptor descriptor, RunBuildBuildOperationType.Details details, OperationStartEvent startEvent) {
+        return new DefaultOperationStartedProgressEvent(startEvent.getStartTime(), descriptor);
+    }
+
+    @Nullable
+    @Override
+    public InternalProgressEvent createProgressEvent(DefaultBuildBuildDescriptor descriptor, OperationProgressEvent progressEvent) {
+        Object details = progressEvent.getDetails();
+        if (details instanceof BuildFailureWithProblemsProgressDetails) {
+            BuildFailureWithProblemsProgressDetails buildFailureDetails = (BuildFailureWithProblemsProgressDetails) details;
+            Throwable failure = buildFailureDetails.getBuildFailure();
+            Map<Throwable, Collection<Problem>> problemsForThrowables = buildFailureDetails.getProblemsForThrowables();
+            InternalFailure rootFailure = DefaultProblemAwareFailure.fromThrowable(failure, problemsForThrowables, ProblemsProgressEventConsumer::createDefaultProblemDetails);
+            return new DefaultBuildFailureEvent(new DefaultBuildFailureDescriptor(operationIdentifierSupplier.get(), descriptor.getId()), new DefaultBuildFailureDetails(Collections.singletonList(rootFailure)));
+        }
+        return null;
```

Comment:
@wolfs What we do here is to create a new progress event just before the root operation event finishes that contains a new `InternalFailure`. 

There's an alternative to consider. We store the problems in a subtype of `InternalFailure`, a type we use for representing failures in finish events anyway. Instead of creating a new event, why don't we add the problems to the finish event (in line 97).

Reply:
The latest (wip) state moved the failures to the finish event. I'm in the process of cleaning up the implementation.